deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/Convolution3D.java

```diff
@@ -0,0 +1,193 @@
+        ret.setIndex(layerIndex);
+        ret.setParamsViewArray(layerParamsView);
+        Map<String, INDArray> paramTable = initializer().init(conf, layerParamsView, initializeParams);
+        ret.setParamTable(paramTable);
+        ret.setConf(conf);
+        return ret;
+    }
+
+    @Override
+    public ParamInitializer initializer() {
+        return Convolution3DParamInitializer.getInstance();
+    }
+
+    @Override
+    public InputType getOutputType(int layerIndex, InputType inputType) {
+        if (inputType == null || inputType.getType() != InputType.Type.CNN3D) {
+            throw new IllegalStateException("Invalid input for Convolution3D layer (layer name=\"" + getLayerName()
+                    + "\"): Expected CNN3D input, got " + inputType);
+        }
+        return InputTypeUtil.getOutputTypeCnn3DLayers(inputType, kernelSize, stride, padding, dilation,
+                convolutionMode, nOut, layerIndex, getLayerName(), Convolution3DLayer.class);
+    }
+
+    @Override
+    public InputPreProcessor getPreProcessorForInputType(InputType inputType) {
+        if (inputType == null) {
+            throw new IllegalStateException("Invalid input for Convolution3D layer (layer name=\"" + getLayerName()
+                    + "\"): input is null");
+        }
+
+        return InputTypeUtil.getPreProcessorForInputTypeCnn3DLayers(inputType, getLayerName());
+    }
+
+
+    @Override
+    public void setNIn(InputType inputType, boolean override) {
+        if (inputType == null || inputType.getType() != InputType.Type.CNN3D) {
+            throw new IllegalStateException("Invalid input for Convolution 3D layer (layer name=\"" + getLayerName()
+                    + "\"): Expected CNN3D input, got " + inputType);
+        }
+
+        if (nIn <= 0 || override) {
+            InputType.InputTypeConvolutional3D c = (InputType.InputTypeConvolutional3D) inputType;
+            this.nIn = c.getChannels();
+        }
+    }
+
+    public static class Builder extends ConvolutionLayer.BaseConvBuilder<Builder> {
+
+        private boolean isNCDHW = true; // in libnd4j: 1 - NCDHW, 0 - NDHWC
+        protected int[] kernelSize = new int[] {2, 2, 2};
+        protected int[] stride = new int[] {1, 1, 1};
+        protected int[] padding = new int[] {0, 0, 0};
+        protected int[] dilation = new int[]{0, 0, 0};
+
+
+        public Builder(int[] kernelSize, int[] stride, int[] padding) {
+            super(kernelSize, stride, padding);
+        }
+
+        public Builder(int[] kernelSize, int[] stride) {
+            super(kernelSize, stride);
+        }
+
+        public Builder(int... kernelSize) {
+            super(kernelSize);
+        }
+
+        public Builder() {
+            super(new int[]{1, 1, 1},
+                    new int[]{1, 1, 1},
+                    new int[]{0, 0, 0},
+                    new int[]{0, 0, 0});
+        }
+
+
+        public Builder kernelSize(int... kernelSize) {
```

Comment:
Maybe: add a preconditions check on (a) not null, and (b) length 3 - applies for all methods here.
FYI we have this now: https://github.com/deeplearning4j/nd4j/pull/2845

Also: javadoc on these methods. And make sure you specify the order (h,w,d?) for the dimensions in said javadoc.